### PR TITLE
pybind/rbd: add allow_shrink=True as a parameter to def resize

### PR DESCRIPTION
**Pull request by zy751713126**
Signed-off-by: Zheng Yin zhengyin@cmss.chinamobile.com

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by zy751713126** [comment]
retest this please

File: src/pybind/rbd/rbd.pyx [hunk 1]
```diff
@@ -2096,6 +2098,24 @@ cdef class Image(object):
...
+        old_size = self.size()
+        allow_shrink = True if old_size > size else False
+        cdef:
+            uint64_t _size = size
+            bint _allow_shrink = allow_shrink
+            librbd_progress_fn_t prog_cb = &no_op_progress_callback
+        with nogil:
+            ret = rbd_resize2(self.image, _size, _allow_shrink, prog_cb, NULL)
+        if ret < 0:
+            raise make_ex(ret, 'error resizing (resize2) image %s' % (self.name,))
```
Comment (tchaikov) [inline src/pybind/rbd/rbd.pyx]:
```py
raise make_ex(ret, 'error resizing (resize2) image %s' % self.name)
```
would suffice

Comment (zy751713126) [reply src/pybind/rbd/rbd.pyx]: I fix it, thanks tchaikov

**Comment by zy751713126** [comment]
@dillaman I know rbd_resize has defualt value allow_shrink=true
https://github.com/ceph/ceph/blob/master/src/librbd/librbd.cc#L3371

I add func resize2 has param allow_shrink, Users can fill in the value(true or false) according to their needs.
Sometimes, image resizes that new size must not be less than old size. I think this func resize2
will be useful. thanks a lot.

**Comment by zy751713126** [comment]
@dillaman thanks for your good advice, I have added allow_shrink=True to the existing def resize.

File: src/pybind/rbd/rbd.pyx [hunk 2]
```diff
@@ -303,6 +303,8 @@ cdef extern from "rbd/librbd.h" nogil:
                                  rbd_image_t *image, const char *snap_name)
     int rbd_close(rbd_image_t image)
     int rbd_resize(rbd_image_t image, uint64_t size)
```
Comment (dillaman) [inline src/pybind/rbd/rbd.pyx]: Nit: drop this declaration since it's no longer used

File: src/pybind/rbd/rbd.pyx [hunk 3]
```diff
@@ -2083,18 +2085,28 @@ cdef class Image(object):
...
         :param size: the new size of the image
         :type size: int
+        :param allow_shrink: permit shrinking
+        :type allow_shrink: bool
         """
-        cdef uint64_t _size = size
+        old_size = self.size()
+        if old_size == size:
+            raise InvalidArgument("error new size is equal to old size")
+        if allow_shrink is False and old_size > size:
```
Comment (dillaman) [inline src/pybind/rbd/rbd.pyx]: Nit: ```if not allow_shrink ...```

Comment (zy751713126) [reply src/pybind/rbd/rbd.pyx]: thanks, I will fix it.

File: src/test/pybind/test_rbd.py [hunk 4]
```diff
@@ -444,6 +444,14 @@ def test_resize(self):
         info = self.image.stat()
         check_stat(info, new_size, IMG_ORDER)
 
+    def test_resize_allow_shrink_False(self):
+        new_size = IMG_SIZE * 2
+        self.image.resize(new_size2)
```
Comment (dillaman) [inline src/test/pybind/test_rbd.py]: Undefined variable name ```new_size2```

Comment (zy751713126) [reply src/test/pybind/test_rbd.py]: It is my mistake, thanks.

**Comment by zy751713126** [comment]
retest this please

**Comment by zy751713126** [comment]
retest this please

**Comment by zy751713126** [comment]
retest this please

File: src/pybind/rbd/rbd.pyx [hunk 5]
```diff
@@ -2083,18 +2084,28 @@ cdef class Image(object):
...
 
         :param size: the new size of the image
         :type size: int
+        :param allow_shrink: permit shrinking
+        :type allow_shrink: bool
         """
-        cdef uint64_t _size = size
+        old_size = self.size()
+        if old_size == size:
+            raise InvalidArgument("error new size is equal to old size")
```
[the remark this answers is not in the document]
Comment (zy751713126) [reply src/pybind/rbd/rbd.pyx]: I mean that it doesn't go to with the following operations, when current size is equal to new size. thanks.

File: src/pybind/rbd/rbd.pyx [hunk 6]
```diff
@@ -2083,18 +2084,28 @@ cdef class Image(object):
...
+        if not allow_shrink  and old_size > size:
```
[the remark this answers is not in the document]
Comment (zy751713126) [reply src/pybind/rbd/rbd.pyx]: I will remove it, thanks

File: src/pybind/rbd/rbd.pyx (hunk 5 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/rbd/rbd.pyx]: then why not just early return?

[the remark this answers is not in the document]
Comment (zy751713126) [reply src/pybind/rbd/rbd.pyx]: We know that rbd.pyx is python rbd api in ceph. I think it is a relatively early return. If you know the earlier return, please let me know, thanks a lot.

[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/rbd/rbd.pyx]: sorry for the confusion, what i meant is
```py
if old_size == size:
  return
```

[the remark this answers is not in the document]
Comment (zy751713126) [reply src/pybind/rbd/rbd.pyx]: ok, I see your mean, thanks

**Comment by zy751713126** [comment]
retest this please

**Pull request merged**